lib/awful/rules.lua

```diff
@@ -207,23 +211,169 @@ function rules.matches_list(c, _rules)
+--  add to and overwrite properties in this table
+-- @tparam table callback.callbacks A table of all callbacks scheduled to be
+--  executed after the main properties are applied.
+-- @tparam[opt={}] table depends_on A list of names of sources this source depends on
+--  (sources that must be executed *before* `name`.
+-- @tparam[opt={}] table precede A list of names of sources this source have a
+--  priority over.
+-- @treturn boolean Returns false if a dependency conflict was found.
+function rules.add_rule_source(name, callback, depends_on, precede)
+    depends_on = depends_on  or {}
+    precede    = precede or {}
+    assert(type( depends_on ) == "table")
+    assert(type( precede    ) == "table")
+
+    for _, v in ipairs(rules.rule_sources) do
+        -- Names must be unique
+        assert(
+            v.name ~= name,
+            "Name must be unique, but '" .. name .. "' was already registered."
+        )
+    end
+
+    rules._rule_source_sort:prepend(name, precede    )
+    rules._rule_source_sort:append (name, depends_on )
+
+    local res, err = rules._rule_source_sort:sort()
+
+    if err then
+        gdebug.print_warning("Failed to add the rule source: "..err)
+        return false
+    end
+
     local callbacks = {}
 
-    for _, entry in ipairs(rules.matching_rules(c, rules.rules)) do
-        if entry.properties then
-            for property, value in pairs(entry.properties) do
-                props[property] = value
-            end
+    -- Get all callbacks for *existing* sources.
+    -- It is important to remember that names can be used in the sorting even
+    -- if the source itself doesn't (yet) exists.
+    for _, v in ipairs(rules.rule_sources) do
+        callbacks[v.name] = v.callback
+    end
+
+    rules.rule_sources = {}
+    callbacks[name]    = callback
+
+    for _, v in ipairs(res) do
+        if callbacks[v] then
+            table.insert(rules.rule_sources, 1, {
+                callback = callbacks[v],
+                name     = v
+            })
+        end
+    end
+
+    return true
+end
+
+--- Remove a source.
+-- @tparam string name The source name.
+-- @treturn boolean If the source was removed
+function rules.remove_rule_source(name)
+    for k, v in ipairs(rules.rule_sources) do
+        if v.name == name then
+            table.remove(rules.rule_sources, k)
+            return true
         end
+    end
+
+    return false
+end
+
+-- Add the rules properties
+local function apply_awful_rules(c, props, callbacks)
+    for _, entry in ipairs(rules.matching_rules(c, rules.rules)) do
+        gtable.crush(props,entry.properties or {})
+
         if entry.callback then
             table.insert(callbacks, entry.callback)
         end
     end
+end
+
+--- The default `awful.rules` source.
+--
+-- **Depends on:**
+--
+-- *nothing*
+--
+-- @rulesources awful.rules
+
+rules.add_rule_source("awful.rules", apply_awful_rules, {"awful.spawn"}, {})
```

Comment:
The comment and the code disagrees. This passes `{"awful.spawn"}` for the `depends_on` parameter.
However, I have to admit that I am not quite sure what this means for the priorities... I guess that `awful.spawn` should overwrite rules from `awful.rules`, but then the dependency should be the other way around... Or not? I'm confused by all this wording (the fundamental problem: higher priority callbacks have to be called later so that they are not overwritten, so there is some kind of inversion going on...).

Reply:
I changed the wording. This is why I remove the `after`/`before` names the original version had.